### DISCUSSION TITLE
TypeScript Generics Issue in AlpineComponent

### DISCUSSION THREAD

**Discussion by devskio** [alpinejs/alpine#4071] [7. Bugs (archived)]
The TypeScript setup in AlpineComponent appears to have a compatibility issue with Generics. When attempting to use Generics in an AlpineJs component, the TypeScript compiler throws an error.

**Error message:**
```
TS2322: Type 'Item[]' is not assignable to type 'InferInterceptor<Item>[]'.
  Type 'Item' is not assignable to type 'InferInterceptor<Item>'.
  Type 'ListItem' is not assignable to type 'InferInterceptor<Item>'.
```

**Code example:**
```
import { AlpineComponent } from 'alpinejs';
import axios from "axios";

export interface ListItem {
  content: string;
  uid: number;
}

interface ListXDataContext<Item extends ListItem> {
  items: Array<Item>;
  fetchItems: () => Promise<void>;
}

export type ListOptions = {
  getItemsUrl: string;
}

const List = <Item extends ListItem>(options: ListOptions):AlpineComponent<ListXDataContext<Item>> => {
  return {
    items: [],

    init() {
      this.fetchItems();
    },

    fetchItems() {
      let items:Array<Item> = [];

      return axios<Array<Item>>({
        url: options.getItemsUrl
      })
        .then((response) => {
          if (response.data.length) {
            items = response.data;
          }
        })
        .catch((error) => {
          console.error('Oops an error occurred: ', error);
        }).finally(() => {
          this.items = items;
        });
    }
  }
}

export default List;
```

**Steps to Reproduce:**
1. Create a TypeScript AlpineJs component using Generics.
2. Attempt to compile the code using TypeScript.
3. Observe the TypeScript compiler error mentioned above.

**Expected Behavior:**
The TypeScript compiler should not throw an error when using Generics in AlpineJs components.

**Additional Information:**
AlpineJs Version: 3.13.5
TypeScript Version: 5.3.3

**Comment by ekwoka** [comment]
hmm...

that's strange.

Technically this repo doesn't include the types for that, as I maintain those on DefinitelyTyped. I haven't run into this personally.

I'll get on this quickly :)

**Reply by ekwoka** [reply]
Initial investigation: the cause is from the fact it can't be inferred that Item is NOT an interceptor, since Item could extend ListItem while also being an Interceptor.

This may be tricky to figure out...

Does the item need to be recognized as being extending ListItem here, or could it just be ListItem?

**Reply by devskio** [reply]
Hello, thanks for the answer.

The idea behind was to have some kind of "abstract" List component that could be extended. 
Imagine you have a list of news, list of events, list of products,... and for each you would like to either use same component, but with different `ListItem` type, or to have completely new Component that would "extend" the Abstract one.

```
export interface NewsListItem extends ListItem {
  date: Date
}

const NewsList = (options: ListOptions):ListComponent => {
  return {
    ...List<NewsListItem>(options),
    orderItems() {
      this.items = this.items.sort((a, b) => {
         // sort functionality
      }
    }
  }
}

**Reply by ekwoka** [reply]
Would `List` then need to return an `AlpineComponent`? Couldn't it just return a normal object?
